src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -351,6 +358,16 @@ protected virtual void ClearProjectData(ProjectId projectId)
         protected internal virtual void ClearDocumentData(DocumentId documentId)
             => this.ClearOpenDocument(documentId);
 
+        /// <summary>
+        /// Allows workspaces to register themselves to be considered when LSP requests come in. Any workspace
+        /// registered will be probed for a matching document/solution which can be given to the request handler
+        /// to operate on.
+        /// </summary>
+        protected void RegisterForLsp()
+        {
+            _lspWorkspaceRegistrationService.Register(this);
+        }
```

Comment:
Adding dependency on LSP to workspace layer is odd. LSP is higher-level concept than workspace.

Reply:
I believe the registration belongs to the host layer (VS e.g.). 

---
In reply to: [533051848](https://github.com/dotnet/roslyn/pull/49661#discussion_r533051848) [](ancestors = 533051848)

Reply:
I don't know how to achieve that, but still have a `Register` method that is callable from specific workspaces. `IEventListener` looks like its based on `WorkspaceKind` which doesn't work for this. eg, at the moment we want to only register VisualStudioWorkspace, MiscellaneousFileWorkspace, MetadataAsSourceWorkspace and AnyCodeWorkspace.

Reply:
I don't see how it would easily be at the VS layer either if we have workspaces we want included (vs, misc, etc) and workspaces we don't (preview).

If this is needed to be moved to IEventListener perhaps an idea would be to set a property on the workspace that says if it's participating in LSP (defaulted to true in the base workspace) that can be overriden (e.g. by preview workspace).   to false).  In the listener this property could be checked since it should have the workspace instance available and do the registration action.

Reply:
Opt-in sounds scary to me, but either way, having a property on Workspace, or a Register method, seem like the same thing, as far as "lsp shouldn't be in workspaces".

Is there some other way to tag a workspace? If workspaces could have multiple kinds then we could add `WorkspaceKind.LSPVisible` or something, but that seems like a huge change

Reply:
Why not check for kinds `Host`, `MiscellaneousFiles` and `MetadataAsSource`?

Reply:
That defeats the purpose of this job though, making the workspaces opt in. eg, hopefully this can be used in future by AnyCode for their workspace, as it is currently broken (they just use reflection to try to kick off diagnostics)

Reply:
I don't think this is the right approach. Workspace layer should not depend on LSP. We need to figure out something else.

Reply:
Why couldn't AnyCodeWorkspace opt-in by implementing some MEF service or calling some APIs, if they want to use LSP?

Reply:
That was the original thought, with each workspace importing ILspWorkspaceRegistrationService and calling Register on it, but that seemed like it wasn't what people wanted. Or do you mean something different?